Support for mixed precision (closes issue #1277)

rupertford opened the pull request:
This PR takes the datatype and precision information provided from the algorithm layer and uses it to determine the precision used in the LFRic PSy-layer. At the same time it adds checks for consistency between the metadata descriptions of arguments and their declarations in the algorithm layer. This check has already found a number of mistakes in our test examples.

rupertford commented:
We now raise an exception when a) the algorithm layer datatype is inconsistent with the kernel metadata or b) the algorithm layer datatype information can't be found (in the cases when it is required). The problem is that this exception is raised as the ``invoke, schedule, kernel, argument`` tree is being built and therefore not all information has been initialised. I have reordered some constructor statements so we can get access to the kernel and module name, however, it is not possible (without a lot of restructuring) to get information about the invoke (algorithm name, name of calling routine etc), so the errors messages can't provide this unfortunately. We get the name of the variable and the name of the kernel (code) which is OK, but it could be better.

rupertford commented:
Chris has a solution to the problem of operators being declared with a base class. This solution is on LFRic ticket #2765. It will require a change to the PSyclone code generation (declarations in the PSy-layer) and he has kindly agreed to create a separate issue to document this.

rupertford commented:
Here are all the exceptions I currently get with a recent version of lfric ...

```
gungho/advection/

rk_transport_rho_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'flux_coeffs' in kernel 'poly1d_flux_coeffs_code'.

rk_transport_theta_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'w1_multiplicity' in kernel 'sample_field_code'.

gungho/iterative_solvers/

dense_operator_alg_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'y_vector' in kernel 'matrix_vector_code'.

diagonal_preconditioner_alg_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'y_vector' in kernel 'x_divideby_y'.

mass_matrix_operator_alg_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'y_vector' in kernel 'setval_c'.

mixed_operator_alg_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'x_vector' in kernel 'combine_w2_field_code'.

mixed_schur_preconditioner_alg_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'y_vector' in kernel 'setval_c'.

multigrid_preconditioner_alg_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'self_r_mg_vector' in kernel 'inc_ax_plus_y'.

pressure_diag_precon_alg_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'y_vector' in kernel 'dg_matrix_vector_code'.

pressure_operator_alg_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'y_vector' in kernel 'apply_helmholtz_operator_code'.

pressure_precon_alg_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'y_vector' in kernel 'tri_solve_code'.

semi_implicit_solver_alg_mod.x90
Generation Error: It was not possible to determine the field type from the algorithm layer for argument 'vector_state_vector' in kernel 'combine_w2_field_code'.

socrates/

illuminate_alg_mod.x90
Generation Error: Scalars must have their precision defined in the algorithm layer but 'timestep' in 'illuminate_code' does not.

radiation_alg_mod.x90
Generation Error: Scalars must have their precision defined in the algorithm layer but 'timestep' in 'sw_code' does not.
```

TeranIvy commented:
@rupertford, which version of the trunk are you working with? I fixed these errors in [LFRic ticket #2737](https://code.metoffice.gov.uk/trac/lfric/ticket/2737) that was merged to trunk 10 days ago.

TeranIvy commented:
@rupertford, which version of the LFRic trunk are you working with? I fixed at least the scalar errors in [LFRic ticket #2737](https://code.metoffice.gov.uk/trac/lfric/ticket/2737) that was merged to trunk 10 days ago.

rupertford commented:
Thanks @TeranIvy. Yes, this version of LFRic is before you put on your changes. We can ignore those particular exceptions.

rupertford commented:
@christophermaynard, this branch seems to be working correctly now. I've tested it with the latest version of LFRic and it does not raise any exceptions. I've not compiled or run the LFRic code but the tests produce the code I would expect so I'm reasonably confident. You can therefore test using this branch as all I now need to do is complete the tests, documentation and examples before putting it forward for review.

rupertford commented:
@TeranIvy has reported a problem with this branch (many thanks). It appears that we are only adding r_def when a scalar, field or operator that requires it is found. If none-of these are found, but r_def is required due to e.g. quadrature, then this is currently not added. Below is her summary ...

```
[22/09 15:05] Kavcic, Iva (Guest)
    
I think #1349 has a bug of not generating "r_def" when it is required. Compilation of one of the generated Psy-layer modules fails with 



algorithm/update_prognostic_scalars_alg_mod_psy.f90(62): error #6683: A kind type parameter must be a compile-time constant.   [R_DEF]
      REAL(KIND=r_def), allocatable :: basis_w3_qr(:,:,:,:), basis_wtheta_qr(:,:,:,:), basis_aspc2_chi_qr(:,:,:,:), &
----------------^
algorithm/update_prognostic_scalars_alg_mod_psy.f90(65): error #6683: A kind type parameter must be a compile-time constant.   [R_DEF]
      REAL(KIND=r_def), pointer :: weights_xy_qr(:) => null(), weights_z_qr(:) => null()
----------------^


Edited​[22/09 15:10] Kavcic, Iva (Guest)
    
This is the branch I am testing with the PSyclone environment built from #1349: https://code.metoffice.gov.uk/trac/lfric/log/LFRic/branches/dev/ivakavcic/r31352_TestPrec. The algorithm that creates the incorrect PSy-layer code is gungho/source/algorithm/update_prognostic_scalars_alg_mod.x90 . 

[22/09 15:30] Kavcic, Iva (Guest)

    I think I know where the problem is. In #1349 it seems that use constants_mod in the generated PSy layer adds precisions based on scalar, field and operator types but not also on the accompanying machinery such as quadrature and evaluator types - this is where the code fails.

​[22/09 15:32] Kavcic, Iva (Guest)
    
Remembering a chat with Chris, we don't plan to change quadrature precisions and they will stay r_def. LFRicConstants.QUADRATURE_TYPE_MAP can be updated with intrinsic type and kind and kind can be added to use constants_mod in the relevant class.

```

TeranIvy commented:
I rebuilt the PSyclone tests environments from the updated branch. Unfortunately, @stevemullerworth reported another failure.

_The updated version is not working for me at all. It fails to use ``r_def`` from constants_mod if an algorithm calls only built-ins with literal ``r_def`` args. e.g.:_
```
algorithm/iterative_solvers/mass_matrix_operator_alg_mod_psy.f90:53:43:

         y_vector_proxy%data(df) = 0.0_r_def
                                           1
Error: Missing kind-parameter at (1)
```
``Ironically, adding a dummy built-in that passed in an r_def literal was the workaround for the previous bug.``

See https://code.metoffice.gov.uk/trac/lfric/ticket/2864 for comments.

TeranIvy commented:
I also have a design comment on precisions of quadrature objects. I see that they are now set from the default kinds in configuration. Quadratures are defined objects in their own right, with the specified intrinsic type and precision (see e.g. https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/trunk/infrastructure/source/field/quadrature_xoyoz_mod.F90). It would make more sense to extend ``LFRicConstants.QUADRATURE_TYPE_MAP`` with "intrinsic" and "precision" properties and pull the information from there.

rupertford commented:
OK, I think this is ready for first review now. Probably best if it is done by @arporter

arporter commented on src/psyclone/parse/algorithm.py:
```diff
@@ -753,19 +774,42 @@ def get_kernel(parse_tree, alg_filename, arg_type_defns):
+                    elif candidate_datatype != datatype:
+                        raise NotImplementedError(
+                            f"Found two non-matching literals within an "
+                            f"expression passed into an invoke from the "
```
"....expression ({str(argument)}) passed into an..."

rupertford replied:
Better, thanks.

arporter commented on src/psyclone/parse/algorithm.py:
```diff
@@ -753,19 +774,42 @@ def get_kernel(parse_tree, alg_filename, arg_type_defns):
+                arguments.append(Arg('literal', argument.tostr().lower(),
+                                 datatype=datatype))
             else:
                 raise NotImplementedError(
```
You could do `if walk(argument, Name): then raise NotImplementedError` first and then you could reduce the indentation of the code above.

rupertford replied:
Good point.

arporter commented on src/psyclone/parse/kernel.py:
```diff
@@ -557,9 +557,10 @@ def get_procedure(ast, name, modast):
+                # We support:
                 # PROCEDURE, nopass :: code => <proc_name> or
                 # PROCEDURE, nopass :: <proc_name>
+                # or nothing, if there is an explicit interface
```
Although this is true, in this branch of the `if` we do only support the two things listed. You could move this whole comment out of the if and replace it with a shorter one.

arporter replied:
Or, you could just revert this file :-)

rupertford replied:
I removed it.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -1456,9 +1458,8 @@ class CodedKern(Kern):
+    :param bool check: whether to check for consistency between the \
+        kernel metadata and the algorithm layer. Defaults to True.
 
     :raises GenerationError: if(check) and the number of arguments in the \
```
This routine no longer raises this exception.

rupertford replied:
Oops.

rupertford commented:
@TeranIvy has kindly done some testing of this branch and has found that the `infer_datatype` method is not producing the required mixed precision results for scalars. I have taken a look and it seems that the method will also not work for fields and operators as well. Mixed precision with `infer_datatype` is obviously not being tested in this PR and should be.

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -388,6 +388,204 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+appropriate subroutine based on the precision of its argument(s).
+
+Below is a simple example of an algorithm code calling the same
+generic kernel twice with potentially different precision and for the
```
Modified.

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -388,6 +388,204 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+
+It is not mandatory for PSyclone to be able to determine the datatype
+of a scalar from the algorithm layer. This constraint was considered
+to be too restrictive as PSyclone only looks at the current algorithm
```
You my English improve thank you.

arporter replied:
Well, apart from the repeated "currently" but I'm going to let myself off :-)

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -388,6 +388,204 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+If the precision information for a scalar is found by PSyclone then
+this is used. If the scalar declaration is found and it contains no
+precision information then PSyclone will abort with a message that
+indicates the problem. If no precision information is found then
```
Yes, that it a very good clarification/justification.

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -388,6 +388,204 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+default precision values are used as specified in the PSyclone config
+file (`r_def` for real, `i_def` for integer and `l_def` for logical).
+
+Supported precisions are `r_def` and `r_solver` for real data, `i_def`
```
Well we are in the scalars section :-) but I've added it for clarity (to the scalars section that talks about scalars and includes information about precision for scalars).

rupertford commented on examples/lfric/README.md:
```diff
@@ -360,6 +360,12 @@ psyclone advection_alg_mod.x90
 
+## Example 19: Mixed precision
+
+This example shows the use of the LFRic mixed precision support to
```
Updated.

arporter replied:
Not really :-) I meant at line 365 - the section heading is fine as it was.

rupertford replied:
Done now :-)

rupertford commented on examples/lfric/eg11/helmholtz_solver_alg_mod.x90:
```diff
@@ -47,12 +47,16 @@ module helmholtz_solver_alg_mod
     type(field_type), intent(inout) :: Hp
     type(field_type), intent(in)    :: p
+    type(field_type) :: grad_p, hb_inv, u_normalisation
+    type(operator_type) :: div_star
```
OK, it now compiles. I think you will have something to say about the Makefile though :-)

rupertford commented on examples/lfric/eg19/algorithm.x90:
```diff
@@ -0,0 +1,55 @@
+! -----------------------------------------------------------------------------
+! Author: R. W. Ford, STFC Daresbury Laboratory
+
+! Example of calling the same kernel with data in the algorithm layer
```
Updated thanks.

rupertford commented on examples/lfric/eg19/algorithm.x90:
```diff
@@ -0,0 +1,55 @@
+program mixed_precision
+
+  use constants_mod, only : r_def, r_solver
+  use mixed_mod, only : mixed_type
```
OK

rupertford commented on examples/lfric/eg19/mixed_mod.f90:
```diff
@@ -0,0 +1,64 @@
+
+  implicit none
+
+  type, extends(kernel_type) :: mixed_type
```
OK

rupertford commented on src/psyclone/dynamo0p1.py:
```diff
@@ -422,7 +422,7 @@ class DynKernelArguments(Arguments):
         adds no additional functionality to its base class other than
         ensuring that initialisation is performed correctly. '''
-    def __init__(self, call, parent_call):
+    def __init__(self, call, parent_call, check=True):
```
Given that this class does not document any of its arguments and - as you say - we will be deleting this API soon, I will leave it as it is. I think that is what you were suggesting.

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -388,6 +388,204 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+
+In order to support mixed precision, PSyclone needs to know the
+precision (as specified in the algorithm layer) of any LFRic data that
+has a type that supports different multiple precisions. The reason for
```
Much nicer thanks.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2257,9 +2273,16 @@ def _invoke_declarations(self, parent):
-        parent.add(DeclGen(parent, datatype="real",
-                           kind=api_config.default_kind["real"],
+        # There is no need to check whether the default kind has been
+        # declared as it as it always will be by this point.
```
Updated. I've also removed the words about not checking as we actually do check in subsequent lines, the first of which is `const_mod = const.UTILITIES_MOD_MAP["constants"]["module"]`

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2257,9 +2273,16 @@ def _invoke_declarations(self, parent):
+        const_mod = const.UTILITIES_MOD_MAP["constants"]["module"]
+        const_mod_list = self._invoke.invokes.psy. \
+            infrastructure_modules[const_mod]
+        if my_kind not in const_mod_list:
```
It is :-)

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -2800,21 +2823,25 @@ def _invoke_declarations(self, parent):
+                field_datatype_map[(arg.data_type, arg.module_name)] = [arg]
+
+        # Add the Invoke subroutine argument declarations for the
+        # different fields types
```
Added, thanks.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3069,47 +3096,60 @@ def _invoke_declarations(self, parent):
+                field_datatype_map[
+                    (arg.proxy_data_type, arg.module_name)] = [arg]
+
+        # Add the Invoke subroutine argument declarations for the
```
Modified.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3378,17 +3418,42 @@ def _create_declarations(self, parent):
+                infrastructure_modules[const_mod]
         # Real scalar arguments
         for intent in FORTRAN_INTENT_NAMES:
             if self._real_scalars[intent]:
```
Phew :-)

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3471,16 +3540,24 @@ def _invoke_declarations(self, parent):
+                                   entity_decls=operators_names,
                                    intent="in"))
+            op_mod = operators_list[0].module_name
             (self._invoke.invokes.psy.infrastructure_modules[op_mod].
```
Added

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3624,6 +3701,13 @@ def _invoke_declarations(self, parent):
+            const_mod = const.UTILITIES_MOD_MAP["constants"]["module"]
+            const_mod_list = self._invoke.invokes.psy. \
+                infrastructure_modules[const_mod]
+            if cma_kind not in const_mod_list:
```
Added

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4498,10 +4586,16 @@ def initialise(self, parent):
+            const_mod = const.UTILITIES_MOD_MAP["constants"]["module"]
+            const_mod_list = self._invoke.invokes.psy. \
+                infrastructure_modules[const_mod]
+            if my_kind not in const_mod_list:
```
Added.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4714,10 +4814,19 @@ def _initialise_xyoz_qr(self, parent):
+            const_mod = const.UTILITIES_MOD_MAP["constants"]["module"]
+            const_mod_list = self._invoke.invokes.psy. \
+                infrastructure_modules[const_mod]
+            if kind not in const_mod_list:
```
Done blah blah blah

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4790,10 +4899,17 @@ def _initialise_face_or_edge_qr(self, parent, qr_type):
+            const_mod = const.UTILITIES_MOD_MAP["constants"]["module"]
+            const_mod_list = self._invoke.invokes.psy. \
+                infrastructure_modules[const_mod]
+            if kind not in const_mod_list:
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -7453,7 +7569,7 @@ def _setup_basis(self, kmetadata):
                 break
 
-    def _setup(self, ktype, module_name, args, parent):
+    def _setup(self, ktype, module_name, args, parent, check=True):
```
Oops

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -7764,12 +7880,10 @@ def gen_stub(self):
             entities(self).declarations(sub_stub)
 
-        # Add "use" statement with kinds (precisions) of all arguments
+        # Add "use" statement with supported argument kinds (precisions)
```
Yes, collecting kinds for stub generation was tricky for some reason and I kept getting missing declaration errors for different cases so I thought it simpler to just wildcard in this case. It's only a temporary solution before we move to using the PSyIR so I thought it was OK.

arporter replied:
It's fine with me. I was always surprised you were taking the trouble to list them explicitly :-)

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8632,74 +8750,208 @@ def ref_name(self, function_space=None):
         supported for scalar, field and operator arguments.
 
-        '''
+        :param alg_datatype_info: the type and precision of this \
```
Oops, I think that was a cut-n-paste error.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8632,74 +8750,208 @@ def ref_name(self, function_space=None):
+            # the algorithm information is available and is not being ignored.
+            if use_alg_info and alg_datatype and \
+               alg_datatype != self.intrinsic_type:
+                raise GenerationError(
```
:-)

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8632,74 +8750,208 @@ def ref_name(self, function_space=None):
+            if use_alg_info and alg_datatype and not alg_precision and \
+               not self.is_literal:
+                raise GenerationError(
+                    f"Scalars must have their precision defined in the "
```
Justifying is good, thanks.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8632,74 +8750,208 @@ def ref_name(self, function_space=None):
+                # is expected that they should match the precision of
+                # the field they are reducing. At the moment there is
+                # an assumption that the precision will always be a
+                # particular specified value.
```
To be honest I'm not sure as it was done a long time ago. I've created an issue to capture this assumption/constraint.

rupertford commented on src/psyclone/gocean0p1.py:
```diff
@@ -349,7 +349,7 @@ class GOKern(CodedKern):
         metadata. Uses this information to generate appropriate PSy layer
         code for the Kernel instance. Specialises the gen_code method to
         create the appropriate GOcean specific kernel call. '''
```
I'm going to argue against this. There is currently no documentation of arguments for the associated classes and we are going to remove the gocean0.1 API in the near future.

arporter replied:
Yep, fair point(s).

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8632,74 +8750,208 @@ def ref_name(self, function_space=None):
         # All supported operators have the same intrinsic type, 'real', so we
         # check by their argument type
         if self.is_operator:
 
```
Done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8632,74 +8750,208 @@ def ref_name(self, function_space=None):
 
-        # All supported scalars have the same metadata, 'GH_SCALAR', so we
-        # check by their intrinsic type
         if self.is_scalar:
```
Done.

rupertford commented:
I've fixed the broken compilation tests by a) adding r_solver* to the PSyclone LFRic infrastructure and b) fixing an invalid kernel that seems to be a pre-existing error.

There is one test (which runs 4 times) that still fails. This is for the intrinsic real_x which uses the precision symbol, but this is not being declared. When fixing this I should also test that r_solver also works.

rupertford commented:
I think this is ready for next review from @arporter

arporter commented on examples/lfric/eg11/Makefile:
```diff
@@ -31,16 +31,60 @@
 
-compile: transform
-	@echo "No compilation supported for lfric/eg11"
+EXEC = compiles
```
This is a very strange name for a binary! Please change to erm, "wont_run.exe" or something.

rupertford replied:
Changed to will_not_run.exe - a very strange name for an executable :-P

arporter commented on examples/lfric/eg11/Makefile:
```diff
@@ -31,16 +31,60 @@
+#          J. Henrichs, Bureau of Meteorology
+# Modifications: R. W. Ford, STFC Daresbury Lab
+
+# The compiler to use may be specified via the F90 environment
```
"...environment variable."

rupertford replied:
Modified.

arporter commented on examples/lfric/eg11/Makefile:
```diff
@@ -31,16 +31,60 @@
+
+F90FLAGS += -I$(LFRIC_PATH)
 
 run: compile
```
Please add a comment explaining why it won't run.

rupertford replied:
Done.

arporter commented on examples/lfric/eg11/Makefile:
```diff
@@ -31,16 +31,60 @@
+
+transform: helmholtz_solver_alg_psy.f90 helmholtz_solver_alg_mod.f90
+
+helmholtz_solver_alg_psy.f90: helmholtz_solver_alg_mod.x90
```
I think you can combine this and line 69:

    helmholtz_solver_alg_psy.f90 helmholtz_solver_alg_mod.f90: helmholtzblahblah.x90
          ${PSYCLONE} -s ...... -opsy .... -oalg .... $<

rupertford replied:
I knew you wouldn't like that. I quite like separating out the dependencies. Anyway, the reviewer knows best.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4825,6 +4832,9 @@ def _initialise_xyoz_qr(self, parent):
                 infrastructure_modules[const_mod]
             if kind not in const_mod_list:
+                # Record that we will need to import the kind for a
+                # declaration (associated with quadrature) from the
```
s/quadrature/basis functions/ ?

rupertford replied:
I think this is associated with quadrature. We add the precision (kind) associated with `gh_quadrature_xyoz`.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2464,7 +2461,23 @@ def __init__(self, call, arg_info, access):
+        # purposes. This may be updated when arg_init1() is called.
+        self._name = self._orig_name
+
+    def arg_init1(self, arg_info):
```
Please rename this (maybe "init_arg_symbol"?) Could it also be private?

rupertford replied:
As mentioned earlier I've renamed it, and pre-pended an underscore.

arporter commented on src/psyclone/parse/algorithm.py:
```diff
@@ -771,40 +771,39 @@ def get_kernel(parse_tree, alg_filename, arg_type_defns):
                     f"variables are not yet supported '{type(argument)}', "
                     f"value '{str(argument)}', kernel '{parse_tree}' in "
                     f"file '{alg_filename}'.")
+            # This is a literal so store the full expression as a
```
Please re-format this comment.

arporter commented on src/psyclone/tests/domain/lfric/lfric_builtins_test.py:
```diff
@@ -2881,6 +2892,26 @@ def test_int_X(tmpdir, monkeypatch, annexed, dist_mem):
+    scalar with precision that is not the default i.e. not i_def. At
+    the moment there is no other integer precision so we make one
+    up. However, this does mean that we can't check whether it
+    compiles.
```
I would have expected PSyclone to reject an unrecognised precision. Are there rules about precision in the LFRic API - i.e. is there a specified list of precision parameters that must be used?

rupertford replied:
The value is monkeypatched in to avoid any tests. I've tried to make that clear in the text.

arporter commented on src/psyclone/tests/domain/lfric/lfric_builtins_test.py:
```diff
@@ -2881,6 +2892,26 @@ def test_int_X(tmpdir, monkeypatch, annexed, dist_mem):
+    kern = first_invoke.schedule.children[0].loop_body[0]
+    monkeypatch.setattr(kern.args[0], "_precision", "i_solver")
+    code = str(psy.gen)
+    assert ("USE constants_mod, ONLY: i_solver, i_def" in code)
```
pylint complains that the parentheses are unnecessary on this line and the next.

rupertford replied:
Oops.

arporter commented on src/psyclone/tests/domain/lfric/lfric_integer_builtins_test.py:
```diff
@@ -1110,3 +1110,23 @@ def test_real_X(tmpdir, monkeypatch, annexed, dist_mem):
+    kern = first_invoke.schedule.children[0].loop_body[0]
+    monkeypatch.setattr(kern.args[0], "_precision", "r_solver")
+    code = str(psy.gen)
+    assert ("USE constants_mod, ONLY: r_solver, i_def" in code)
```
pylint isn't going to like these parentheses (and on the next line too).

rupertford replied:
Oops.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1813,7 +1823,7 @@ class for an r_solver_operator.
     assert ("It was not possible to determine the operator type from the "
             "algorithm layer for argument 'self_mm_w0' in kernel "
-            "'testkern_operator_code'.")
+            "'testkern_operator_code'." in str(info.value))
```
Oops! Well spotted.

arporter commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1566,6 +1555,26 @@ def test_arg_intrinsic_type_error():
+    assert ("Supported argument types are scalar, field and operator, but "
+            "the argument 'a' in kernel 'testkern_code' is none of these."
+            in str(info.value))
+    
```
pylint complains about trailing whitespace here.

rupertford replied:
Oops.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/constants_mod.F90:
```diff
@@ -54,7 +55,21 @@ module constants_mod
+
+
   integer, parameter :: real_type    = 1 !< A parameter used to indicate a real data typa
+  integer, parameter :: r_solver_real_type = 1 !< A parameter used to indicate a r_solver data type
```
Does it matter that this has the same value as `real_type`? Or have you just copied this out of the LFRic infrastructure?

rupertford replied:
It's copied. I just checked and can confirm that the values are the same in the LFRic infrastructure.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/testkern_ref_elem_mp_mod.f90:
```diff
@@ -63,11 +63,7 @@ subroutine testkern_ref_elem_mp_code(nlayers, fld1, ndf_w1, undf_w1, map_w1, &
-    integer(kind=i_def), intent(in) :: ndf_w3
-    integer(kind=i_def), intent(in), dimension(ndf_w3) :: map_w3
-    integer(kind=i_def), intent(in) :: undf_w1, undf_w2, undf_w3
+    integer(kind=i_def), intent(in) :: undf_w1
```
Please update the copyright date too.

rupertford replied:
Done.

arporter reviewed (changes requested):
Looking pretty good now.
I think it's all minor things to fix.
Examples and tests are now all OK with compilation.
I've checked with pylint - see inline.
All docs build fine.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8681,12 +8697,15 @@ def __init__(self, kernel_args, arg_meta_data, arg_info, call, check=True):
+        # Complete the initialisation of the argument (after
+        # _init_data_type_properties() so the precision info etc is
+        # already set up)
+        self.arg_init1(arg_info)
```
I've called it `_complete_init()`. Hopefully you will think this is a better name!

rupertford commented on src/psyclone/gocean0p1.py:
```diff
@@ -415,6 +415,9 @@ class GOKernelArgument(KernelArgument):
         KernelArgument.__init__(self, arg, arg_info, call)
+        # Complete the argument initialisation as in some APIs it
+        # needs to be separated.
+        self.arg_init1(arg_info)
```
Changed them all.

rupertford commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -1696,15 +1705,26 @@ def gen_code(self, parent):
         field_name2 = self.array_ref(self._arguments.args[0].proxy_name)
         field_name1 = self.array_ref(self._arguments.args[1].proxy_name)
+        precision = self._arguments.args[0].precision
         rhs_expr = ("real(" + field_name1 + ", " +
```
Updated (but kept the precision variable for clarity). Hopefully that is OK.

rupertford commented:
Ready for next review from @arporter

arporter reviewed (changes requested):
I realised that I still hadn't looked at all the test files (and there are a lot!).
Nothing major but I do have some questions about why particular tests have been removed.
I also gave up requesting that the copyrigh
def test_dynamopsy_inf_modules():
    '''Check that the infrastructure_modules() method of DynamoPSy behaves
    as expected.
    '''
    dynamo_psy = DynamoPSy(DummyInvokeInfo())
    assert (dynamo_psy.infrastructure_modules is
            dynamo_psy._infrastructure_modules)t date in each file be updated so please could you do that for all modified files.
All tests and examples are OK with compilation.
I've tried to highlight some conversations that remain ongoing from previous reviews.

rupertford commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -406,6 +408,54 @@ def get_scalar_argument_references(self):
+
+        :param parent: Node in f2pygen tree to which to add call.
+        :type parent: :py:class:`psyclone.f2pygen.BaseGen`
+
```
Oops.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/infrastructure/constants_mod.F90:
```diff
@@ -6,6 +6,38 @@
 ! LICENCE.original is available from the Met Office Science Repository Service:
 ! https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/trunk/LICENCE.original
 !-----------------------------------------------------------------------------
+! BSD 3-Clause License
```
I added the r_solver constant and any associated declarations associated with mixed precision for this PR. Perhaps you are asking whether I copied this additional code from the LFRic infrasrtucture. If so, yes I did so it is not my code, but the file is not the same as the version on LFRic trunk as I only copied what was needed. Perhaps I should keep the Met Office copyright and simply say I've modified it? Or just copy the complete latest version across?

arporter replied:
Ah I see. As you have modified it (rather than copied it unchanged) then I'm happy with it as is.

rupertford commented on tutorial/practicals/LFRic/single_node/code/helmholtz_solver_alg_mod.x90:
```diff
@@ -66,12 +66,12 @@ contains
                                     ptheta2 => null(), &
                                     p3theta => null()
     real(kind=r_def)             :: timestep_term
     type(field_type),    pointer :: mt_lumped_inv => null()
```
Done (and updated for all modified files)

rupertford commented on tutorial/practicals/LFRic/single_node/1_openmp/helmholtz_solver_alg_mod.x90:
```diff
@@ -66,13 +66,13 @@ contains
                                     p3theta => null()
     real(kind=r_def)             :: timestep_term
     real(kind=r_def)             :: sum
     type(field_type),    pointer :: mt_lumped_inv => null()
```
Done.

rupertford commented on tutorial/practicals/LFRic/distributed_memory/code/helmholtz_solver_alg_mod.x90:
```diff
@@ -66,12 +66,12 @@ contains
     type(field_type),                intent(in)    :: p
```
Done.

rupertford commented on tutorial/practicals/LFRic/distributed_memory/code/helmholtz_solver_alg_mod.x90:
```diff
@@ -66,12 +66,12 @@ contains
     type(field_type),    pointer :: mt_lumped_inv => null()
+    type(field_type) :: div_star
 
     call invoke( setval_c(grad_p, 0.0_r_def),                                &
```
Added before the invoke.

rupertford commented on tutorial/practicals/LFRic/distributed_memory/4_reductions/helmholtz_solver_alg_mod.x90:
```diff
@@ -66,12 +66,12 @@ contains
     real(kind=r_def)             :: timestep_term
     type(field_type),    pointer :: mt_lumped_inv => null()
+    type(field_type) :: div_star
 
```
Added before the invoke.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/testkern_operator_real_mod.f90:
```diff
@@ -0,0 +1,88 @@
+! OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Author: R. W. Ford, STFC Daresbury Lab
+
```
We have a kernel that is similar but passes an integer scalar. I needed a real scalar to check for different (r_solver) precision.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/mixed_mod.f90:
```diff
@@ -0,0 +1,64 @@
+module mixed_mod
```
Updated. I think the filename/module names are following the convention i.e. `x_mod.f90` where `x` is `mixed` in this case. Am I missing something?

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/4.8_multikernel_invokes.f90:
```diff
@@ -47,7 +47,7 @@ program multikernel_invokes_7
   implicit none
 
-  type(field_type)           :: a, b, c, d, e(3), f, g
+  type(field_type)           :: a, b, c, d, e(3), f, g, h
```
Done.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/4.4_multikernel_invokes.f90:
```diff
@@ -37,7 +37,7 @@ program multikernel_invokes_5
   ! operator information
 
-  use constants_mod,         only : r_def
+  use constants_mod,         only : i_def
```
Done.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/4.16_multikernel_invokes_real_int_scalar_invalid.f90:
```diff
@@ -36,17 +36,22 @@
 program multikernel_invokes_real_int_scalar_invalid
 
```
Done.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/26.8_mixed_precision_args.f90:
```diff
@@ -0,0 +1,55 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2021, Science and Technology Facilities Council.
```
Done.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/26.6_mixed_precision_solver_vector.f90:
```diff
@@ -29,32 +29,44 @@
-! Modified: I. Kavcic, Met Office
+! Author: R. W. Ford STFC Daresbury Lab
+!
+! Example where the field is dereferenced from an abstract_vector_type
```
I don't think so. I think the original test no longer worked as PSyclone could not work out what the precision was so I repurposed it to cover inferring the datatype when a field is in an abstract vector type (it must be a field type according to LFRic restrictions).

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/26.2_mixed_precision_self.f90:
```diff
@@ -41,7 +41,7 @@ module self_example
   use r_solver_operator_mod, only : r_solver_operator_type
   use quadrature_xyoz_mod, only : quadrature_xyoz_type
-  use testkern_operator_mod, only : testkern_operator_type
+  use testkern_operator_real_mod, only : testkern_operator_type
```
Done.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/26.1_mixed_precision.f90:
```diff
@@ -31,29 +31,23 @@
 ! -----------------------------------------------------------------------------
 ! Author: R. W. Ford, STFC Daresbury Lab
 !
```
Done.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/23.5_ref_elem_mixed_prec.f90:
```diff
@@ -0,0 +1,50 @@
+! -----------------------------------------------------------------------------
+! BSD 3-Clause License
+!
+! Copyright (c) 2021, Science and Technology Facilities Council.
```
Done.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/20.5_multi_cma_invoke.f90:
```diff
@@ -1,7 +1,7 @@
 ! BSD 3-Clause License
 !
-! Copyright (c) 2017-2018, Science and Technology Facilities Council
+! Copyright (c) 2017-2021, Science and Technology Facilities Council
```
Done.

rupertford commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -1014,47 +1014,6 @@ def test_call_multi_reduction_error(monkeypatch, dist_mem):
         "or builtin" in str(err.value))
 
 
-def test_reduction_no_set_precision(monkeypatch, dist_mem):
```
This change seems to have gone due to another pull request.

rupertford commented on src/psyclone/tests/lfric_ref_elem_test.py:
```diff
@@ -201,6 +202,7 @@ def test_refelem_arglist_err():
 
 
 def test_refelem_gen(tmpdir):
+
```
Oops

rupertford commented on src/psyclone/tests/lfric_ref_elem_test.py:
```diff
@@ -327,3 +329,20 @@ def test_all_faces_refelem_gen(tmpdir):
+def test_refelem_no_rdef(tmpdir):
+    '''Check that the PSy-layer declares r_def if there is a reference
+    element specified in the metadata but the arguments are not of
+    type r_def. This is rquired as array arguments associated with the
```
Oops.

rupertford commented on src/psyclone/tests/dynamo0p3_lma_test.py:
```diff
@@ -691,69 +697,6 @@ def test_operator_nofield_scalar(tmpdir):
             "weights_xy_qr, weights_z_qr)" in gen)
 
 
-def test_operator_nofield_scalar_deref(tmpdir, dist_mem):
```
These tests seem to have already been removed in a different PR

rupertford commented on src/psyclone/tests/domain/lfric/lfric_dynamopsy_test.py:
```diff
@@ -0,0 +1,169 @@
+
+def test_dynamopsy_gen_no_invoke():
+    '''Check that the gen() method of DynamoPSy behaves as expected for a
+    minimal psy-layer with no invoke calls.
```
I've reworded. I meant, when the algorithm layer does not have any invoke calls.

rupertford commented on src/psyclone/tests/domain/lfric/lfric_dynamopsy_test.py:
```diff
@@ -0,0 +1,169 @@
+
+    '''
+    dynamo_psy = DynamoPSy(DummyInvokeInfo())
+    assert (dynamo_psy.infrastructure_modules is
```
It's a great test. I've updated the docstring to explain what it is doing.

rupertford commented on src/psyclone/tests/domain/lfric/lfric_dynamopsy_test.py:
```diff
@@ -0,0 +1,169 @@
+
+def test_dynamopsy_gen():
+    '''Check that the gen() method of DynamoPSy behaves as expected for a
+    psy-layer with invoke calls. Simply check the invoke calls are
```
Re-wordified.

rupertford commented on src/psyclone/tests/alggen_test.py:
```diff
@@ -269,40 +269,10 @@ def test_multi_deref_derived_type_args():
     print(gen)
     assert (
         "CALL invoke_0(f1, obj_a % iflag, f2, m1, m2, obj_b % iflag, "
```
Many of the tests in this file were from the time when we used to add print to the tests as they then gave information if the test went wrong. However, as we don't do that anymore I've removed all of them.

rupertford commented on src/psyclone/tests/alggen_test.py:
```diff
@@ -269,40 +269,10 @@ def test_multi_deref_derived_type_args():
         in gen)
 
 
-def test_op_and_scalar_and_qr_derived_type_args():
```
They cause exceptions as the datatype can't be found so are invalid tests. The former is still used in another context (in dynamo0p3_lma_test.py:714) and the latter has been re-used for an exception test in dynamo0p3_test.py:1941

rupertford commented:
Ready for next review from @arporter

rupertford commented:
Taking this back to fix the problem brought up by Tom about the implicit assumption that abstract vector types are r_def precision (which they might not be in Tom's code). The proposed PSyclone solution is to raise an exception here saying that we don't know what the type is (we special case abstract vector types anyway so it should simply be a case of removing the special case). The proposed solution for the user is to add in a pointer to the correct concrete type, point it at the abstract field inside the appropriate select type clause and then pass the pointer into the invoke. Tom has reported that this works for him and Chris is already using this in other situations. This problem is captured in issue #1614.

rupertford commented:
Ready for next review from @arporter. I've fixed the bug identified by Tom i.e. made `abstract_vector_type` raise an exception and added support for `r_solver_field_vector_type` as requested at the previous telco (again by Tom).

arporter commented on .github/workflows/python-package.yml:
```diff
@@ -38,7 +38,8 @@
 # This workflow will install Python dependencies, run tests and lint with a
 # variety of Python versions.
-# For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
+# For more information see:
```
I don't understand why these changes are showing up on this branch?

arporter reviewed (changes requested):
Hi @rupertford, would you mind seeing whether you can sort out this branch so that the 'diff' doesn't show changes that aren't part of this PR?
I'll hold off the next review until you've had a look.

arporter commented on examples/lfric/eg11/Makefile:
```diff
@@ -31,16 +31,63 @@
+helmholtz_solver_alg_psy.f90 helmholtz_solver_alg_mod.f90: helmholtz_solver_alg_mod.x90
+	${PSYCLONE} -s ./async_script.py -opsy helmholtz_solver_alg_psy.f90 -oalg helmholtz_solver_alg_mod.f90 $<
+
+#helmholtz_solver_alg_mod.f90: helmholtz_solver_alg_mod.x90
```
Please rm these two commented lines.

rupertford replied:
Removed.

arporter commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -406,6 +408,54 @@ def get_scalar_argument_references(self):
                 if arg.is_scalar]
 
+
+class LFRicXKern(LFRicBuiltIn):
```
Thanks but you shouldn't have to use the `six` package for this.

rupertford replied:
Yes, quite right. Modified and removed six package from file.

arporter commented on src/psyclone/tests/test_files/dynamo0p3/26.7_mixed_precision_abstract_solver_vector.f90:
```diff
@@ -29,29 +29,44 @@
+! Author: R. W. Ford STFC Daresbury Lab
+!
+! Example where the field is dereferenced from an abstract_vector_type
+! and therefore has no type information. The LFRic rules state that an
```
Is this rule documented somewhere? Just to check that I understand: this is saying nothing about the underlying precision of the field_type?

rupertford replied:
The rule was never documented and was the cause of errors for Tom so is no longer a rule (an exception is now raised). I've updated the description.

arporter reviewed (changes requested):
That was tricky. I struggled to see the right diff this time around.
Tests are all still OK with compilation.
Updated UG builds OK and reads well - just a few minor corrections.
Various minor things need tweaking and you seem to have got overzealous in marking files as modified!

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/1.13_single_invoke_field_deref.f90:
```diff
@@ -40,11 +40,14 @@ program single_invoke
 
   implicit none
 
+  type estate_type
```
This example was failing as the type was not declared in this module. Adding this type allows PSyclone to look within a derived type and find the correct type within it. Updated copyright date.

rupertford commented on src/psyclone/domain/lfric/lfric_builtins.py:
```diff
@@ -408,6 +408,56 @@ def get_scalar_argument_references(self):
+
+        '''
+        # Subclass must set _field_type
+        if not self._field_type:
```
All the PRs associated with mixed precision have been old school gen() and f2pygen. I agree we need to move to the new way. I think Iva hasn't yet done field conversion intrinsics so this could/should be picked up when this is done.

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -384,6 +384,246 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+    real(r_def)    :: x_r_def
+    real(r_solver) :: x_r_solver
+
+    call invoke( example_type(x_r_def),    &
```
Updated.

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -384,6 +384,246 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+Supported field types are `field_type` (which contains `real` data
+with precision `r_def`), `r_solver_field_type` (which contains `real`
+data with precision `r_solver`) and `integer_field_type` (which
+contains `integer` data with precision `i_def`). If an unsupported
```
Good point.

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -384,6 +384,246 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+
+If PSyclone finds an argument that is declared as an
+`abstract_field_type` then it will not know the actual type of the
+argument and will raise an exception::
```
Better thanks.

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -384,6 +384,246 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+    end select
+    ! ...
+
+The suggested solution to this is to add a pointer to the code that is
```
Thanks for the pointer :-)

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -384,6 +384,246 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+argument and passed into the routine::
+
+    ! ...
+    class (abstract_vector_type), intent(inout) :: x
```
Yes I think so. Thanks.

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -384,6 +384,246 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+
+It is not mandatory for PSyclone to be able to determine the datatype
+of a scalar from the algorithm layer. This constraint was considered
+to be too restrictive as PSyclone currently only examines the current
```
Betterer thanks.

rupertford commented on doc/user_guide/dynamo0p3.rst:
```diff
@@ -384,6 +384,246 @@ as arguments and that an Invoke may not mix inter-grid kernels with
+this is used. If the scalar declaration is found and it contains no
+precision information then PSyclone will abort with a message that
+indicates the problem (since this violates LFRic coding standards). If
+no precision information is found then default precision values are
```
Thanks.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3089,47 +3115,60 @@ def _invoke_declarations(self, parent):
+
+        # Add the Invoke subroutine declarations for the different
+        # field-type proxies
+        for fld_type, fld_mod in field_datatype_map:
```
Yes much nicer, although I had to use items() rather than iteritems() as the latter is not supported in Python3.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3089,47 +3115,60 @@ def _invoke_declarations(self, parent):
+                # create new entry
+                operators_datatype_map[op_arg.proxy_data_type] = [op_arg]
+        # Declare the operator proxies
+        for operator_datatype in operators_datatype_map:
```
Yes.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3563,16 +3631,26 @@ def _invoke_declarations(self, parent):
+                # This datatype has not been seen before so create new entry
+                operators_datatype_map[op_arg.data_type] = [op_arg]
+        # Declare the operators
+        for operator_datatype in operators_datatype_map:
```
Changed.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3470,17 +3509,42 @@ def _create_declarations(self, parent):
+                        if real_scalar_kind not in const_mod_list:
+                            const_mod_list.append(real_scalar_kind)
+                    elif self._kernel:
+                        self._kernel.argument_kinds.add(real_scalar_kind)
```
Fixed and added test

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3493,24 +3557,28 @@ def _create_declarations(self, parent):
+                        const_mod_list.append(dkind)
+                elif self._kernel:
+                    self._kernel.argument_kinds.add(dkind)
 
```
As above.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -3493,24 +3557,28 @@ def _create_declarations(self, parent):
+                        const_mod_list.append(dkind)
+                elif self._kernel:
+                    self._kernel.argument_kinds.add(dkind)
 
```
As above.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -7775,12 +7903,15 @@ def _setup(self, ktype, module_name, args, parent):
         :param parent: the parent of this kernel call in the generated \
                        AST (will be a loop object).
         :type parent: :py:class:`psyclone.dynamo0p3.DynLoop`
+        :param bool check: optional argument to do with checking \
```
Copied the suggested text from 8656.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8968,76 +9105,282 @@ def ref_name(self, function_space=None):
-    def _init_data_type_properties(self):
-        '''
-        Set up kernel argument information from LFRicConstants: precision,
+    def _init_data_type_properties(self, arg_info, use_alg_info=True):
```
OK, done.

rupertford commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8968,76 +9105,282 @@ def ref_name(self, function_space=None):
+            information. Optional argument that defaults to True.
+
+        :raises GenerationError: if the datatype for a gh_operator \
+            could not be found in the algorithm layer.
```
Added.

rupertford commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -1015,23 +1015,25 @@ def test_call_multi_reduction_error(monkeypatch, dist_mem):
                      distributed_memory=dist_mem).create(invoke_info)
+
+    # A reduction argument will always have a precision value so we
+    # need to monkeypatch the it.
```
Oops.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/15.21.3_int_a_plus_X_builtin.f90:
```diff
@@ -32,6 +32,7 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Author: I. Kavcic, Met Office
+! Modified: R. W. Ford, STFC Daresbury Lab
```
Very strange. Removed my name.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/15.21.4_int_inc_a_plus_X_builtin.f90:
```diff
@@ -32,6 +32,7 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Author: I. Kavcic, Met Office
+! Modified: R. W. Ford, STFC Daresbury Lab
```
Removed.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/15.22.3_int_a_minus_X_builtin.f90:
```diff
@@ -32,6 +32,7 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Author: I. Kavcic, Met Office
+! Modified: R. W. Ford, STFC Daresbury Lab
```
Removed.

rupertford commented on src/psyclone/tests/test_files/dynamo0p3/15.22.4_int_inc_a_minus_X_builtin.f90:
```diff
@@ -32,6 +32,7 @@
 ! POSSIBILITY OF SUCH DAMAGE.
 ! -----------------------------------------------------------------------------
 ! Author: I. Kavcic, Met Office
+! Modified: R. W. Ford, STFC Daresbury Lab
```
Removed.

rupertford commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -3833,3 +4296,46 @@ def test_read_only_fields_hex(tmpdir):
     assert expected in generated_code
+
+
+def test_mixed_precision_args():
```
The answer is yes it can be checked but no it does not work. Our copy of the LFRic infrastructure needs to be updated to include `r_solver_operator_type` and `r_solver_operator_proxy_type`.

rupertford replied:
I've just checked the latest version of lfric_trunk and it does not contain r_solver_operator_type. This functionality must still be on a branch. I have created a new issue #1638, have made the associated test xfail and have referenced the new issue.

rupertford commented on src/psyclone/tests/dynamo0p3_cma_test.py:
```diff
@@ -409,7 +409,7 @@ def test_cma_operator_arg_lfricconst_properties(monkeypatch):
     assert cma_op_arg.proxy_data_type == "columnwise_operator_proxy_type"
     assert cma_op_arg.intrinsic_type == "real"
-    assert cma_op_arg.precision == "r_def"
+    assert cma_op_arg.precision == "r_solver"
```
Updated. The builtin's also look like they need updating but that is separate to this PR.

rupertford commented:
Ready for next review from @arporter. The link problem in the checks seems to be binder refusing connections due to too many users.

arporter reviewed (approved):
All requested changes have been made. All tests and examples are OK.
All docs build fine (and have been updated).
Tom and Iva have checked this branch with LFRic.
